### PR TITLE
jack 1.9.21

### PR DESCRIPTION
**Pull request by HaHeho**
Created with `brew bump-formula-pr`.

### CODE REVIEW

**Comment by nandahkrishna** [comment]
> Is this maybe related to the move of the ["jack example tools"](https://github.com/jackaudio/jack-example-tools) into their own repository?

It seems like `jack_wait` may have moved to that repository. I can take a look at this and attempt a fix in a while.

**Comment by falkTX** [comment]
You can use `--example-tools` to get the old behaviour back, but be advised these tools are only included for this release, they  will be removed in the next one.
See https://jackaudio.org/news/2022/01/15/jack1-v01260-and-jack2-v1920-releases.html for more details

**Comment by nandahkrishna** [comment]
@falkTX I'm facing an issue when building the extra tools locally. On macOS, the internal clients aren't built (`Build internal zita-a2j/j2a client : no`) so I'm seeing this error:
```
../tools/meson.build:83:0: ERROR: Unknown variable "c_args_jack_unload".
```
Would you happen to know how to fix this?

**Comment by falkTX** [comment]
meson version too old? there should be another error before that one I suppose.
@dvzrv might know more, he is the one that setup most of the meson related stuff

**Comment by dvzrv** [comment]
> @falkTX I'm facing an issue when building the extra tools locally. On macOS, the internal clients aren't built (`Build internal zita-a2j/j2a client : no`) so I'm seeing this error:
> 
> ```
> ../tools/meson.build:83:0: ERROR: Unknown variable "c_args_jack_unload".
> ```
> 
> Would you happen to know how to fix this? Hopefully, the CI logs should also be available soon.

@nandahkrishna what is your local jack version and how do you build the jack-example-tools project using meson?
Your issue is in this section of the build script: https://github.com/jackaudio/jack-example-tools/blob/0a23b230a635db7570c4a41eb1b559e34eaee2e2/tools/meson.build#L77-L89
It relies on the detection done here: https://github.com/jackaudio/jack-example-tools/blob/0a23b230a635db7570c4a41eb1b559e34eaee2e2/meson.build#L32-L69 (both seem to result to false)

Please test whether those code snippets for detection contained in the tests compile for you. They have not been tested on macOS, since (at least I) lack a machine to test on (see also https://github.com/jackaudio/jack-example-tools/issues/11).
The snippets rely on an (unconfigured) call to [compiler.compile()](https://mesonbuild.com/Reference-manual_returned_compiler.html#compilercompiles), which may need further locations for headers etc.
If brew provides a default location for pkgconfig integration or headers/libs that we can use, please let me know, then we may add this for the detection of jack1/jack2.

If jack itself is not detected (which I do not believe is the problem here, as you would not get as far as compiling `jack_unload`), you may need to provide further locations to the pkgconfig integration, depending on where your jack is installed, using the `PKG_CONFIG_PATH` variable, e.g.:

```
export PKG_CONFIG_PATH="/my/install/location/of/jack2/lib/pkgconfig:$PKG_CONFIG_PATH"
meson build
ninja -C build
```

File: Formula/jack.rb [hunk 1]
```diff
@@ -39,18 +40,35 @@ class Jack < Formula
...
     system Formula["python@3.10"].opt_bin/"python3", "./waf", "build"
     system Formula["python@3.10"].opt_bin/"python3", "./waf", "install"
 
-    # Remove Python script used to control D-Bus JACK as it isn't enabled in formula
-    rm bin/"jack_control"
+    resource("jack-example-tools").stage do
+      ENV.append_path "PKG_CONFIG_PATH", lib/"pkgconfig"
+      ENV.append_path "LIBRARY_PATH", lib
+
+      inreplace "meson.build", "lib_rt = cc.find_library('rt')", "" if OS.mac?
```
Comment (dvzrv) [inline Formula/jack.rb:64]: @HaHeho Please provide this fix as a pull request towards jack-example-tools!

**Comment by dvzrv** [comment]
FWIW, on a Macbook Pro with Big Sur, that I could use for a while I was able to compile jack-example-tools with:

```
PKG_CONFIG_PATH=/usr/local/Cellar/opus/1.3.1/lib/pkgconfig:/usr/local/Cellar/readline/8.1.2/lib/pkgconfig/ meson build -Dopus_support=disabled
ninja -C build
```

after fixing the lib_rt dependency stuff (will create a PR for it),

after installing (using brew)
* `meson`
* `jack` (1.9.19)
* `pkgconf` (and properly using it for pkg-config: `ln -sv /usr/local/bin/pkgconf pkg-config`)
* `opus` (opus support is disabled, because the brew version does not configure its build with `--enable-custom-modes` and thus does not provide `opus/opus_custom.h`, see https://github.com/archlinux/svntogit-packages/blob/0496258eb6e2d40c0f4bcb3bb48863086bc2660d/trunk/PKGBUILD#L21 -- better detection of whether the required header is around can be implemented in the build system)
* `readline`

File: Formula/jack.rb (hunk 1 above)
Comment (dvzrv) [reply Formula/jack.rb:64]: Nevermind, I will do it :)

**Comment by HaHeho** [comment]
> I am not sure. IMHO it would be better to have jack-example-tools be installable as a separate package (as it makes building against jack probably easier, but that assumption is based on my non-existent knowledge about the brew build stuff...). In general the mentioned PR should allow you to cherry-pick commits to apply to the project to be able to build it without issue though.

From what I understood, the jack-example-tools should be an independent extension to a Jack installation. Ideally, this would work similarly inside the brew ecosystem i.e., jack-example-tools would exist as an independent formula. But I don't know if there is really a need for that (if so, someone could set it up).

For now, the jack-example-tools are included in Jack 1.9.12, but this will be the last release to do so, right? So is it unreasonable to exclude jack-example-tools from building and testing as part of the brew process entirely also in this release already?
If someone explicitly wants them, they will need a different way to get them in the future anyways?

**Comment by dvzrv** [comment]
> > I am not sure. IMHO it would be better to have jack-example-tools be installable as a separate package (as it makes building against jack probably easier, but that assumption is based on my non-existent knowledge about the brew build stuff...). In general the mentioned PR should allow you to cherry-pick commits to apply to the project to be able to build it without issue though.
> 
> From what I understood, the jack-example-tools should be an independent extension to a Jack installation. Ideally, this would work similarly inside the brew ecosystem i.e., jack-example-tools would exist as an independent formula. But I don't know if there is really a need for that (if so, someone could set it up).

Ideally, yes. Especially on Linux we now have three implementations of the JACK API (jack1, jack2 and pipewire-jack - see https://wiki.archlinux.org/title/JACK_Audio_Connection_Kit#Comparison_of_JACK_implementations for a comparison), which led to the decision to split out the jack-example-tools (many setups rely on them and this way people are still able to more or less easily switch between different JACK implementations without too many conflicting files etc.).
If many (or any) people on macOS rely on the libraries and executables provided by jack-example-tools: I don't know.
Either way, I believe it is good to provide them (bundled with jack or as a separate formula).

> For now, the jack-example-tools are included in Jack **1.9.21**, but this will be the last release to do so, right? So is it unreasonable to exclude jack-example-tools from building and testing as part of the brew process entirely also in this release already? If someone explicitly wants them, they will need a different way to get them in the future anyways?

jack2 1.9.21 is the last release where the tools can still be built (the build option is off by default though). After this release (so soonish) they will be removed from the code base and future development will only be done in the context of jack-example-tools.

I guess the inclusion of jack-example-tools (or as a separate formula) should be done rather sooner than later, so it is valid to think about it *now*.

**Comment by nandahkrishna** [comment]
Given the discussion above (and the issues when building jack-example-tools within the `jack` formula), I think it should be okay to create a separate `jack-example-tools` formula. Here's some documentation on how to create a new formula: https://docs.brew.sh/Formula-Cookbook.

You don't need to create a separate PR for the new formula, we can just add it in a separate commit here.

**Comment by nandahkrishna** [comment]
I've modified the `jack` formula to remove jack-example-tools and related tests. It should now be okay to add the new formula (here or in a new PR).

**Comment by nandahkrishna** [comment]
I'll open a separate PR for the jack-example-tools. Thanks everyone!

**Review by nandahkrishna — APPROVED** [review]

**Pull request closed without merging**